### ISSUE TITLE
locations in source proposed move

### ISSUE THREAD

**Issue by ghost** [magefree/mage#2703]
RemoveAllCountersSourceEffect is currently located in mage.abilities.effects.common and probably should be in mage.abilities.effects.common.counter along with all the other add/remove counters effects. I ran across this location difference when seeing if there was existing stuff to use for a card I'm working on.

EDIT: See replies for all proposed moves.

**Comment by JRHerlehy** [comment]
Id say if you do that slight refactor, it might be time to do a bigger reorg and get a tad more explicit on the sorting of abilities and effects in the Mage project.

**Comment by MTGfan** [comment]
I'll keep an eye on it and see what other things could be better organized. I just happened to notice this one was counter related but no in the same location as the rest of the counter effects. Of course if anyone else has anything that would be better suited in another location if those get added to here then when the list is big enough to move forward someone can implement.

**Comment by MTGfan** [comment]
Abilities could be broken down into sub locations. Such as:

mage.abilities.common.activated
mage.abilities.common.static
mage.abilities.common.triggered (may need other locations more specific to the kind of trigger such as combat)

**Comment by Styxo** [comment]
There is also two packages for keywords: abilityword and keyword. Don't know why these are separate.
The framework could definitely use a reorg.

**Comment by MTGfan** [comment]
Ok here is what I have so far for mage.abilities.common the following changes (location name with files to be moved to then:

**mage.abilities.abilityword**

LandfallAbility

**mage.abilities.common.activated**

ActivateAsSorceryActivatedAbility
ActivateIfConditionActivatedAbility
ActivateOnlyByOpponentActivatedAbility
FetchLandActivatedAbility
LimitedTimesPerTurnActivatedAbility
SimpleActivatedAbility
TurnFaceUpAbility
	
**mage.abilities.common.static**

CanBeYourCommanderAbility
CantBeCounteredAbility
ChancellorAbility
SimpleEvasionAbility
SimpleStaticAbility
SkipUntapOptionalAbility

**mage.abilities.common.static.combat**

AttacksEachCombatStaticAbility
AttacksEachTurnStaticAbility
CanBlockOnlyFlyingAbility
CantBlockAbility
DamageAsThoughNotBlockedAbility

**mage.abilities.common.triggered**

AuraAttachedTriggeredAbility
BecomesMonstrousSourceTriggeredAbility
BecomesMonstrousTriggeredAbility
BecomesRenownedSourceTriggeredAbility
BecomesTappedAttachedTriggeredAbility
BecomesTappedSourceTriggeredAbility
BecomesTappedTriggeredAbility
BecomesTargetAttachedTriggeredAbility
BecomesTargetControllerSpellTriggeredAbility
BecomesTargetTriggeredAbility

DiscardedByOpponentTrigger _(this needs a better name that fits what it does such as DiscardSourceOpponentCausedTriggeredAbility)_

DiscardsACardOpponentTriggeredAbility
DrawCardControllerTriggeredAbility
DrawCardOpponentTriggeredAbility
OpponentSacrificesNonTokenPermanentTriggeredAbility
OpponentSacrificesPermanentTriggeredAbility
SacrificeAllTriggeredAbility
SacrificeIfCastAtInstantTimeTriggeredAbility
SacrificeSourceTriggeredAbility
SimpleTriggeredAbility
SpellCastAllTriggeredAbility
SpellCastControllerTriggeredAbility
SpellCastOpponentTriggeredAbility
TapForManaAllTriggeredAbility
TapForManaAllTriggeredManaAbility
TapLandForManaAllTriggeredAbility
TapLandForManaAllTriggeredManaAbility
TurnedFaceUpAllTriggeredAbility
TurnedFaceUpSourceTriggeredAbility
UnattachedTriggeredAbility
ZoneChangeAllTriggeredAbility
ZoneChangeTriggeredAbility

**mage.abilities.common.triggered.battlefield**

AllyEntersBattlefieldTriggeredAbility

AnotherCreatureEntersBattlefieldTriggeredAbility _(probably should rename to EntersBattlefieldAnotherCreatureTriggeredAbility)_

CreatureEntersBattlefieldTriggeredAbility _(probably should rename to EntersBattlefieldCreatureTriggeredAbility)_

EntersAnotherCreatureYourControlTriggeredAbility _(the 4 cards that use this changed accordingly and this can be merged with AnotherCreatureEntersBattlefieldTriggeredAbility by use of targetController in constructor and used as switch in checkTrigger)_

EntersBattlefieldAllTriggeredAbility
EntersBattlefieldControlledTriggeredAbility
EntersBattlefieldTappedAbility
EntersBattlefieldTriggeredAbility
EntersOrLeavesTheBattlefieldSourceTriggeredAbility _(probably could remove the word source)_
LeavesBattlefieldAllTriggeredAbility
LeavesBattlefieldTriggeredAbility

**mage.abilities.common.triggered.graveyard**

PutCardIntoGraveFromAnywhereAllTriggeredAbility _(probably could remove word card)_
PutIntoGraveFromAnywhereSourceTriggeredAbility _(probably could remove the word source)_
PutIntoGraveFromBattlefieldAllTriggeredAbility
PutIntoGraveFromBattlefieldSourceTriggeredAbility _(probably could remove the word source)_

**mage.abilities.common.triggered.step**

BeginningOfCombatTriggeredAbility
BeginningOfDrawTriggeredAbility
BeginningOfEndStepTriggeredAbility
BeginningOfPostCombatMainTriggeredAbility
BeginningOfPreCombatMainTriggeredAbility
BeginningOfUntapTriggeredAbility
BeginningOfUpkeepTriggeredAbility

BeginningOfYourEndStepTriggeredAbility _(redundant as it is already handled by BeginningOfEndStepTriggeredAbility by use of targetController, 19 cards to change accordingly)_

EndOfCombatTriggeredAbility
	
**mage.abilities.common.triggered.combat**

AttackedByCreatureTriggeredAbility
AttacksAllTriggeredAbility
AttacksAloneTriggeredAbility
AttacksAndIsNotBlockedTriggeredAbility
AttacksAttachedTriggeredAbility
AttacksCreatureYouControlTriggeredAbility
AttacksOrBlocksEnchantedTriggeredAbility
AttacksOrBlocksTriggeredAbility
AttacksTriggeredAbility
AttacksWithCreaturesTriggeredAbility
BecomesBlockedAllTriggeredAbility
BecomesBlockedByCreatureTriggeredAbility
BecomesBlockedTriggeredAbility
BlocksAttachedTriggeredAbility
BlocksCreatureAttachedTriggeredAbility
BlocksCreatureTriggeredAbility
BlocksOrBecomesBlockedByCreatureTriggeredAbility
BlocksOrBecomesBlockedTriggeredAbility
BlocksTriggeredAbility
DiesAttachedTriggeredAbility
DiesCreatureTriggeredAbility
DiesThisOrAnotherCreatureTriggeredAbility
DiesTriggeredAbility

**mage.abilities.common.triggered.damage_life**

DamageDealtToAttachedTriggeredAbility
DealsCombatDamageToACreatureTriggeredAbility
DealsCombatDamageToAPlayerTriggeredAbility
DealsCombatDamageTriggeredAbility
DealsDamageAttachedTriggeredAbility
DealsDamageGainLifeSourceTriggeredAbility
DealsDamageToACreatureAttachedTriggeredAbility
DealsDamageToACreatureTriggeredAbility
DealsDamageToAPlayerAllTriggeredAbility
DealsDamageToAPlayerAttachedTriggeredAbility
DealsDamageToAPlayerTriggeredAbility
DealsDamageToOneOrMoreCreaturesTriggeredAbility
DealsDamageToOpponentTriggeredAbility
DealtDamageToSourceTriggeredAbility
GainLifeControllerTriggeredAbility

And this is what is left in **mage.abilities.common**:

AsEntersBattlefieldAbility _(is this incorrectly not a triggered ability? and therefor a redundancy of the enters battlefield triggered ability(ies))_

AttachableToRestrictedAbility

BecomesAuraAttachToManifestSourceEffect _(incorrectly located should be somewhere in mage.effects.common)_

CastCommanderAbility

DiesAndDealtDamageThisTurnTriggeredAbility _(could be in triggered.combat or triggered.damage or somewhere else)_

EntersBattlefieldAbility _(some or all of the cards that use this, do they not have a EntersBattleFieldTriggeredAbility? and thus the constructor in EntersBattlefieldAbility should be added to EntersBattleFieldTriggeredAbility and the cards using EntersBattlefieldAbility changed to use that?)_

EntersBattlefieldOrDiesSourceTriggeredAbility _(could be in triggered.combat or triggered.battlefield or somewhere else)_

ExploitCreatureTriggeredAbility

FeralDeceiverAbility _(two cards use this. It's hard to say if future cards will have the same ability. could use a better name)_

LicidAbility _(this is kind of like ability word abilities but without "licid -" followed by the rule in parenthesis)_

OnEventTriggeredAbility _(every card that uses this except two has at the beginning/end of X step. One of them is just unused import and the other Jihad can be implemented correctly instead of the hack that is using this. Which makes this redundant and it is easier to change the 50 cards using this than the 764 cards that use the individual step triggered abilities)_

PassAbility _(this along with passEffect appear to be AI related and not actual a card ability or effect)_

PayMoreToCastAsThoughtItHadFlashAbility
PlanswalkerEntersWithLoyalityCountersAbility

PutIntoGraveFromAnywhereSourceAbility _(correct me if I'm wrong but the cards that use this their abilities are the same as PutIntoGraveFromAnywhereSourceTriggeredAbility just variation in wording If/When meaning the same. If this is not a redundancy probably could remove the word source)_

**Comment by JRHerlehy** [comment]
So I'll ask, why is everything in the "common" package.  That seems like an unnecessary package level if they fit into one of the subpackages.  And if "common" is being used as the catchall for ones that don't fit in a single category, maybe it needs a bit of a rename as well.

And I'm personally not a fan of underscores in package names.  That to me signifies that the package could be split up into two separate packages.

There is a lot of other cleanup that needs to be done.  In most projects, the packages start at mage.  In some, they start at org.mage, in another, it starts at com.xmage, and in another, it is com.magefree.  For one project that is maintained by a single group, having their published packages start all over the place is a bit of a mess.

These are just my opinions about the matter thought.  I'm sure there are reasons in the past that they are organized as they are.

**Comment by MTGfan** [comment]
If you look at how effects are organized you will understand why I compiled the list of proposed moves as you requested additional stuff in need of organization. All those that I listed are all lumped into one location currently. And makes it a real pain in the rear end for anyone looking for something they probably need to use for a card but don't know the exact thing but the location organization narrows down what to look through and what not. Obviously as the game has grown everyone just kept making new classes and putting them in the catch all location mage.abilities.common has become.
